src/rascal/math/hyp1f1.hh

```diff
@@ -30,487 +30,185 @@
 
 #include "rascal/math/utils.hh"
 
-#include <map>
 #include <vector>
 
 namespace rascal {
   namespace math {
-
-    // This is kept as a header only for a small but not marginal performance
-    // gain
-
-    /**
-     * Utilities to apply the recurrence relations of 1F1 from maximal
-     * values of a and b to minimal values.
-     * The recurrence path is chosen so that derivatives of 1F1 are also
-     * a result.
-     *
-     * G refers to the modified 1F1 defined below
-     */
-    inline double recurence_to_val_downward(double a, double b, double z,
-                                            double M1p2p, double M1p1p) {
-      return z * (a - b) * M1p2p / (b * (b + 1)) + M1p1p;
-    }
-
-    inline double recurence_to_der_downward(double a, double b, double z,
-                                            double M2p3p, double M1p2p) {
-      return z * (a + 1) * M2p3p / ((b + 2) * (b + 1)) + M1p2p;
-    }
-
-    inline double recurence_G_to_val_downward(double a, double b, double z,
-                                              double M1p2p, double M1p1p) {
-      return (z * (a - b) * M1p2p + M1p1p * b) / a;
-    }
-
-    inline double recurence_G_to_der_downward(double, double b, double z,
-                                              double M2p3p, double M1p2p) {
-      return z * M2p3p + M1p2p * (b + 1);
-    }
-
-    /**
-     * Computes the 1F1 with the direct sum
-     *  @f[
-     *      1F1(a,b,z) = \sum_{j=0}^{\infty} \frac{(a)_j}{(b)_jj!} z^{j}
-     *  @f]
-     *
-     *  @f[
-     *      G(a,b,z) = \frac{\Gamma(a)}{\Gamma(b)} * \exp{-\alpha r_{ij}^2}
-     *                    * 1F1(a,b,z)
-     *  @f]
-     */
-    class Hyp1f1Series {
-     protected:
-      double a, b;
-      size_t mmax;
-      double prefac;
-      double tolerance;
-      bool is_exp{false};
-
-      Eigen::VectorXd coeff{};
-      Eigen::VectorXd coeff_derivative{};
-
-     public:
-      // Parameter used to fix the sum when evaluating the numerical
-      // derivatives.
-      int n_terms{100};
-
-      Hyp1f1Series(double a, double b, size_t mmax, double tolerance = 1e-14)
-          : a{a}, b{b}, mmax{mmax}, prefac{std::tgamma(a) / std::tgamma(b)},
-            tolerance{tolerance} {
-        // when a == b, 1F1 is an exponential
-        if (std::abs(1 - this->b / this->a) < DBL_FTOL) {
-          this->is_exp = true;
-        } else {
-          coeff.resize(mmax);
-          coeff_derivative.resize(mmax);
-          double u{0.};
-          // precomputes the (a)_i / (b)_i / i! coefficients
-          coeff(0) = a / b;
-          coeff_derivative(0) = (a + 1) / (b + 1);
-          for (size_t i{1}; i < mmax; ++i) {
-            u = (a + i) / ((i + 1) * (b + i));
-            coeff(i) = coeff(i - 1) * u;
-            u = (a + i + 1) / ((i + 1) * (b + i + 1));
-            coeff_derivative(i) = coeff_derivative(i - 1) * u;
-          }
-        }
-      }
+    namespace internal {
```

Comment:
Why the move to `internal`? That namespace was only meant for things that don't need to be seen outside that file (or outside a group of two or three files). Furthermore, none of the other maths utilities are in this namespace, so this move is going to be confusing for everyone.

Reply:
Because these classes are not used outside of this file. The only 'external' class here is `Hyp1f1SphericalExpansion`

Reply:
>  Furthermore, none of the other maths utilities are in this namespace, so this move is going to be confusing for everyone.

For utilities only used inside `rascal::math`, I think they should be moved to `internal` too then. For utilities used outside of `rascal::math`, they should stay in there.

Reply:
But the reason I said this was confusing is because these are high-level maths utilities, like calculating the hypergeometric function or modified spherical Bessel or whatever, that should be visible in the `rascal::math` namespace (which is already sequestered away from the main rascal namespace, so the rationale for having another `internal` to prevent namespace pollution is even weaker). The “main” classes for each file (like Hyp1f1 and its SphericalExpansion variant here) should be visible in `rascal::math`, if anything to aid discoverability in case someone wants to implement a new representation and would like to see what math functions we already implement.

It’s fine to put “internal” classes, like `Hyp1f1Series` and `Hyp1f1Asymptotic`, into `internal` because they’re not *meant* to be used outside the math utils. But the main classes, which are meant to be useful outside `rascal::math`, should stay as part of the math interface.

Reply:
> should be visible in rascal::math, if anything to aid discoverability in case someone wants to implement a new representation and would like to see what math functions we already implement.

I don't see how namespaces aid or prevent discoverability, but I agree anyway that `Hyp1f1` could also live in `rascal::math`. I mainly wanted to move `Hyp1f1Series` and `Hyp1f1Asymptotic` to `internal`, and `Hyp1f1` followed since it was not used by itself. I'll move it back.

Reply:
As an example for discoverability: Let’s say we decide to make the API documentation more manageable by excluding anything within namespaces called `internal` (which would certainly be a reasonable suggestion). But this would then remove anything inside `rascal::math::internal` from the public documentation, perhaps leading a new developer to believe that we don’t have a dedicated Hyp1f1 class (just the specialization for spherical expansion). Of course, once they go into the code they’d find it, but we don’t want to make people search the source code for what is effectively API documentation.

Reply:
updated

Reply:
Where? Looks like `Hyp1f1` is still in `namespace internal`.

Reply:
Nope, see https://github.com/cosmo-epfl/librascal/pull/194/files#diff-18b2af0cf0528f03643f85559ceaaa0dR145 (GH getting confused by force-pushes again?)

Reply:
Ah, I was looking at an old local copy. Github isn't the only one getting confused by all this rewriting of history...